input: stdin: Describe EoF behaviour, parser

ringerc opened the pull request:
Document how to use a parser with the stdin plugin to read arbitrary messages, since the default behaviour is to assume a specific json input format. This is very confusing for new users, since the plugin just logs

    invalid record found, it's not a JSON map or array

when it sees inputs it doesn't expect, and the docs don't explain that you can set a parser to handle other formats.

Document the stdin plugin's behaviour on end-of-file.

Also explain the accepted input formats for the plugin's default reader: object, object with timestamp, or object with timestamp and metadata.

ringerc commented:
How do I progress this?

patrick-stephens commented:
What do you need? I can review.

ringerc commented:
@patrick-stephens Review would be greatly appreciated, yes. Thankyou.

patrick-stephens commented on pipeline/inputs/standard-input.md:
```diff
@@ -1,18 +1,37 @@
 # Standard Input
 
-The **stdin** plugin allows to retrieve valid JSON text messages over the standard input interface \(stdin\). In order to use it, specify the plugin name as the input, e.g:
+The **stdin** plugin allows to retrieve a message stream from the standard input interface \(stdin\) of the fluent-bit process. In order to use it, specify the plugin name as the input, e.g:
```
Let's ensure we use `Fluent Bit` everywhere and split sentences to simplify reviews.

```suggestion
The **stdin** plugin supports retrieving a message stream from the standard input interface \(stdin\) of the Fluent Bit process.
In order to use it, specify the plugin name as the input, e.g:
```

patrick-stephens commented on pipeline/inputs/standard-input.md:
```diff
@@ -1,18 +1,37 @@
+
+## Input formats
+
+If no parser is configured for the stdin plugin, it expects *valid json* input input data with one of the following data formats:
```
```suggestion
If no parser is configured for the stdin plugin, it expects *valid json* input data with one of the following data formats:
```

I'm not sure those are valid JSON though - the `=>` is something `stdout` uses?

ringerc replied:
@patrick-stephens The output here isn't json because the example uses the default format of the `stdout` plugin. The _input_ to the `stdin` plugin is required to be valid json unless a parser is configured explicitly.

patrick-stephens commented on pipeline/inputs/standard-input.md:
````diff
@@ -1,18 +1,37 @@
 2. [ time, { map => val, map => val, map => val } ]
 ```
 
+It will log errors like
````
I'm a little confused here - above it says to expect data in this format then it says there will be an error.

Can we clarify with correct format and then examples of failing ones?

ringerc replied:
@patrick-stephens I can see how it would've been confusing.

* here's what you should provide
* here's the error you get if you violate that expectation

I'll rearrange.

ringerc replied:
Rephrased to

---

If no parser is configured for the stdin plugin, it expects *valid json* input input data with one of the following data formats:
 
```bash
1. { map => val, map => val, map => val }
2. [ time, { map => val, map => val, map => val } ]                                           
```

Any input data that is *not* in one of the above formats will cause the plugin to log errors like: 
   
```
[error] [input:stdin:stdin.0] invalid record found, it's not a JSON map or array
```
 
To handle inputs in other formats, a parser must be explicitly specified in the configuration for the `stdin` plugin. See [parser input example](#parser-input-example) for sample configuration.

patrick-stephens commented on pipeline/inputs/standard-input.md:
````diff
@@ -48,11 +67,56 @@ Fluent Bit v1.x.x
+$
+```
+
+In most cases you'll want to use a more realistic parser that splits your
````
Try not to use shortenings of words so `you'll` should be `you will`. I would also try to avoid any kind of personal pronoun like `you`.

ringerc replied:
Reworded to

> In real-world deployments it is best to use a more realistic parser that splits               
> messages into real fields and adds appropriate tags.

ringerc commented:
@patrick-stephens I've updated this docs PR. I addressed your comments re clarity. 

I've also updated the `concepts/key-concepts.md` doc to reflect the updates made in 9aa10d2bdc0913f94dd366f24b64ef539f0d4b20 since that PR did not update the docs. I wanted to link to the fluent-bit event input format from the docs for the stdin plugin, but found that those docs were outdated so I had to update them.

patrick-stephens commented on concepts/key-concepts.md:
```diff
@@ -30,12 +30,36 @@ Jan 18 12:52:16 flb gsd-media-keys[2640]: # watch_fast: "/org/gnome/terminal/leg
+
+### Event format
+
+The Fluent Bit wire protocol represents an Event is a 2-element array
```
```suggestion
The Fluent Bit wire protocol represents an Event as a 2-element array
```

patrick-stephens commented on concepts/key-concepts.md:
```diff
@@ -30,12 +30,36 @@ Jan 18 12:52:16 flb gsd-media-keys[2640]: # watch_fast: "/org/gnome/terminal/leg
+* METADATA is a possibly-empty object containing event metadata; and
+* MESSAGE is an object containing the event body.
+
+Fluent-bit versions prior to v2.1.0 instead used:
```
My pet peeve, let's be consistent on "Fluent Bit":

```suggestion
Fluent Bit versions prior to v2.1.0 instead used:
```

ringerc commented:
@patrick-stephens I applied both your changes. Thankyou.

ringerc commented:
@patrick-stephens Thankyou, and much appreciated.
